File: books/OOPC/src/exercises/ch02/new.c
#include "new.h"
#include "new.r"

#include <stdlib.h>
#include <assert.h>

void *new(const void *_class, ...) {
    const struct Class *class = _class;
    void *p = calloc(1, class->size);

    assert(p);

    /**
     * p --> object --> class
     */
    *(const struct Class **)p = class;

    if (class->ctor) {
        va_list ap;
        va_start(ap, _class);
        p = class->ctor(p, &ap);
        va_end(ap);
    }
    return p;
}

void delete(void *self) {
    const struct Class **cp = self;

    if (self && *cp && (*cp)->dtor) {
        self = (*cp)->dtor(self);
    }
    free(self);
}

File: books/OOPC/src/exercises/ch02/new.r
/**
 * Type-descriptor
 */
#ifndef CLASS_H
#define CLASS_H

#include <stddef.h>
#include <stdarg.h>

struct Class {
    size_t size;
    void *(*ctor)(void *self, va_list *app);
    void *(*dtor)(void *self);
    void *(*clone)(const void *self);
    int (*differ)(const void *self, const void *b);
};

#endif
